### PR TITLE
SONARJAVA-1050 RSPEC-2097 Raise issue only if there is a cast of the parameter

### PR DESCRIPTION
**Pull request by merciesa**

### CODE REVIEW

**Comment by SonarTech** [comment]
SonarQube analysis reported no new issues.

File: java-checks/src/main/java/org/sonar/java/checks/EqualsArgumentTypeCheck.java [hunk 1]
```diff
@@ -72,15 +73,53 @@ public void visitNode(Tree tree) {
...
+    @Override
+    public void visitTypeCast(TypeCastTree tree) {
+      if (isArgument(tree.expression())) {
+        hasCast = true;
+      } else {
+        super.visitTypeCast(tree);
+      }
+    }
+
+    private boolean isArgument(ExpressionTree tree) {
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/EqualsArgumentTypeCheck.java]: This method is duplicated, maybe create a static version with an expression tree and the symbol as parameter.

File: java-checks/src/main/java/org/sonar/java/checks/EqualsArgumentTypeCheck.java [hunk 2]
```diff
@@ -112,7 +151,11 @@ public void visitBinaryExpression(BinaryExpressionTree tree) {
     @Override
     public void visitMethodInvocation(MethodInvocationTree tree) {
       if (EQUALS_MATCHER.matches(tree)) {
-        typeChecked = true;
+        if (tree.methodSelect().is(Tree.Kind.MEMBER_SELECT) && isArgument(((MemberSelectExpressionTree) tree.methodSelect()).expression())) {
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/EqualsArgumentTypeCheck.java]: this line does exactly the same as block 175-178. A method can probably be extracted!

File: java-checks/src/main/java/org/sonar/java/checks/EqualsArgumentTypeCheck.java [hunk 3]
```diff
@@ -112,46 +152,36 @@ public void visitBinaryExpression(BinaryExpressionTree tree) {
...
     }
 
     private boolean isGetClassOnArgument(ExpressionTree tree) {
       ExpressionTree expressionTree = removeParenthesis(tree);
       if (expressionTree.is(Tree.Kind.METHOD_INVOCATION) && GETCLASS_MATCHER.matches((MethodInvocationTree) expressionTree)) {
-        ExpressionTree methodSelect = ((MethodInvocationTree) expressionTree).methodSelect();
-        if (methodSelect.is(Tree.Kind.MEMBER_SELECT)) {
-          ExpressionTree expression = ((MemberSelectExpressionTree) methodSelect).expression();
-          return isArgument(expression);
+        if (isInvocationOnArgument((MethodInvocationTree) expressionTree)) {
```
Comment (SonarTech) [inline java-checks/src/main/java/org/sonar/java/checks/EqualsArgumentTypeCheck.java]: :new: MAJORMerge this if statement with the enclosing one.[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1066)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 new issue:
* 1 major

Details:
* [Merge this if statement with the enclosing one.](https://github.com/SonarSource/sonar-java/blob/327deb19bf80b6fcef9e64836c6bd1b601361994/java-checks/src/main/java/org/sonar/java/checks/EqualsArgumentTypeCheck.java#L170)[](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1066)

**Comment by SonarTech** [comment]
SonarQube analysis reported no new issues.

**Comment by m-g-sonar** [comment]
LGTM!

**Comment by merciesa** [comment]
Reopened

**Comment by SonarTech** [comment]
SonarQube analysis reported no new issues.

File: java-checks/src/test/files/checks/EqualsArgumentType.java [hunk 4]
```diff
@@ -1,231 +1,224 @@
...
 
   public boolean equals(String that) { // Compliant, different signature
+    return ((Object) that) == null;
   }
 
   @Override
-  public boolean equals(Object that) { // Noncompliant {{Add a type test to this method.}}
+  public boolean equals(Object that) { // Compliant
+    if (true) {
+      return (Object) this;
```
Comment (m-g-sonar) [inline java-checks/src/test/files/checks/EqualsArgumentType.java]: Use another example of expression than a return statement... That example is hard to follow.

File: java-checks/src/main/java/org/sonar/java/checks/EqualsArgumentTypeCheck.java [hunk 5]
```diff
@@ -111,47 +173,32 @@ public void visitBinaryExpression(BinaryExpressionTree tree) {
...
-      ExpressionTree expressionTree = removeParenthesis(tree);
-      return expressionTree.is(Tree.Kind.IDENTIFIER) && ((IdentifierTree) expressionTree).symbol().equals(parameterSymbol);
-    }
-
-    private boolean isExplicitComparison(ExpressionTree operand1, ExpressionTree operand2) {
-      return (isArgument(operand1) && isThis(operand2)) || (isArgument(operand2) && isThis(operand1));
-    }
-
-    private boolean isGetClassOnArgument(ExpressionTree tree) {
+    private boolean isSuperOrThis(ExpressionTree tree) {
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/EqualsArgumentTypeCheck.java]: What about `this.foo.bar()` ?

Comment (merciesa) [inline java-checks/src/main/java/org/sonar/java/checks/EqualsArgumentTypeCheck.java]: Could indeed happen in some kind of proxy object.

Changed so that a method call type checking if it takes the argument as input:
this.bar(that) // Compliant
this.foo.bar(null, that) // Compliant

this.bar() // Noncompliant, that is not referenced
this.foo.bar() // Noncompliant, that is not referenced

**Comment by m-g-sonar** [comment]
LGTM!
